// @workadventure/scripting-api-extra/dist/VariablesExtra.js
import { Properties } from "./Properties";
export class VariableDescriptor {
    constructor(object) {
        this.name = object.name;
        this.x = object.x;
        this.y = object.y;
        this.properties = new Properties(object.properties);
    }
    get isReadable() {
        const readableBy = this.properties.getString("readableBy");
        if (!readableBy) {
            return true;
        }
        return WA.player.tags.includes(readableBy);
    }
    get isWritable() {
        const writableBy = this.properties.getString("writableBy");
        if (!writableBy) {
            return true;
        }
        return WA.player.tags.includes(writableBy);
    }
}
export async function getAllVariables() {
    const map = await WA.room.getTiledMap();
    const variables = new Map();
    getAllVariablesRecursive(map.layers, variables);
    return variables;
}
function getAllVariablesRecursive(layers, variables) {
    for (const layer of layers) {
        if (layer.type === "objectgroup") {
            for (const object of layer.objects) {
                if (object.type === "variable") {
                    variables.set(object.name, new VariableDescriptor(object));
                }
            }
        }
        else if (layer.type === "group") {
            getAllVariablesRecursive(layer.layers, variables);
        }
    }
}
